File: Attacked.ts
import Board from "./Board.js";
import Pieces from "./Pieces.js";

namespace Attacked {
    export var attacked: Set<number> = new Set();

    export function getAttacked(): void {
        attacked.clear();
    
        for (let index of Board.pieces[!Board.turn ? 1 : 0]) {
            // console.log(index);
            let piece: string = Pieces.PIECES[Board.board[index]].toLowerCase();

            switch (piece) {
                case "q":
                    rook(index);
                    bishop(index);
                    break;
                case "r":
                    rook(index);
                    break;
                case "b":
                    bishop(index);
                    break;
                case "n":
                    knight(index);
                    break;
                case "p":
                    pawn(index);
                    break;
                case "k":
                    king(index);
                    break;
            }
        }
    }

    export function pawn(loc: number): void {
        for (let offset of Pieces.Offsets.PAWN_CAPTURE[!Board.turn ? 1 : 0]) {
            let i: number = loc + offset;
            if (Board.board[i] == -1)
                continue;
            attacked.add(i);
        }
    }

    export function knight(loc: number): void {
        for (let offset of Pieces.Offsets.KNIGHT) {
            let i: number = loc + offset;
            if (Board.board[i] == -1)
                continue;

            attacked.add(i);
        }
    }

    export function bishop(loc: number): void {
        let opponentKing: number = Pieces.PIECES.indexOf(Board.turn ? "k" : "K");
        for (let offset of Pieces.Offsets.BISHOP) {
            let i: number = loc + offset;
            while (Board.board[i] != -1) {
                attacked.add(i);

                if (Board.board[i] && Board.board[i] != opponentKing)
                    break;

                i += offset;
            }
        }
    }

    export function rook(loc: number): void {
        let opponentKing: number = Pieces.PIECES.indexOf(Board.turn ? "k" : "K");
        for (let offset of Pieces.Offsets.ROOK) {
            let i: number = loc + offset;
            while (Board.board[i] != -1) {
                attacked.add(i);

                if (Board.board[i] && Board.board[i] != opponentKing)
                    break;

                i += offset;
            }
        }
    }

    export function king(loc: number): void {
        for (let offset of Pieces.Offsets.KING) {
            let i: number = loc + offset;
            if (Board.board[i] == -1)
                continue;

            attacked.add(i);
        }
    }
}

export default Attacked;

File: attacked.cpp
#include <vector>
#include <unordered_set>

#include "attacked.h"
#include "board.h"
#include "checks.h"
#include "constants.h"

void attacked::getAttacked(std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    for (int index : board::pieces[!board::turn]) {
        char piece = tolower(board::board[index]);

        switch (piece) {
            case 'q': {
                attacked::rchecks(index, attacked, attackedSet);
                attacked::bchecks(index, attacked, attackedSet);
                break;
            }

            case 'r': {
                attacked::rchecks(index, attacked, attackedSet);
                break;
            }

            case 'b': {
                attacked::bchecks(index, attacked, attackedSet);
                break;
            }

            case 'n': {
                attacked::nchecks(index, attacked, attackedSet);
                break;
            }

            case 'p': {
                attacked::pchecks(index, attacked, attackedSet);
                break;
            }

            case 'k': {
                attacked::kchecks(index, attacked, attackedSet);
                break;
            }
        }
    }
}
std::vector<checks::Check> attacked::getAttackedForJS() {
    std::vector<checks::Check> attacked;
    std::unordered_set<int> attackedSet;
    attacked::getAttacked(&attacked, &attackedSet);
    return attacked;
}

void attacked::pchecks(int loc, std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    for (int offset : constants::PAWN_CAPTURE[!board::turn]) {
        int i = loc + offset;
        if (board::board[i] == -1)
            continue;

        attacked->push_back({ i, loc, -10 });
        attackedSet->insert(i);
    }
}
void attacked::nchecks(int loc, std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    for (int offset : constants::KNIGHT) {
        int i = loc + offset;
        if (board::board[i] == -1)
            continue;

        attacked->push_back({ i, loc, -10 });
        attackedSet->insert(i);
    }
}
void attacked::bchecks(int loc, std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    char opponentKing = board::turn ? 'k' : 'K';
    for (int offset : constants::BISHOP) {
        int i = loc + offset;
        while (board::board[i] != -1) {
            attacked->push_back({ i, loc, -offset });
            attackedSet->insert(i);

            if (board::board[i] && board::board[i] != opponentKing)
                break;

            i += offset;
        }
    }
}
void attacked::rchecks(int loc, std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    char opponentKing = board::turn ? 'k' : 'K';
    for (int offset : constants::ROOK) {
        int i = loc + offset;
        while (board::board[i] != -1) {
            attacked->push_back({ i, loc, -offset });
            attackedSet->insert(i);

            if (board::board[i] && board::board[i] != opponentKing)
                break;

            i += offset;
        }
    }
}
void attacked::kchecks(int loc, std::vector<checks::Check>* attacked, std::unordered_set<int>* attackedSet) {
    for (int offset : constants::KING) {
        int i = loc + offset;
        if (board::board[i] == -1)
            continue;

        attacked->push_back({ i, loc, -10 });
        attackedSet->insert(i);
    }
}